documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md

```diff
@@ -0,0 +1,277 @@
+# 2023-06-01 Implementation Plan: Nuxt 3 Migration
+
+**Author**: @obulat
+
+<!-- See the implementation plan guide for more information: https://github.com/WordPress/openverse/tree/19791f51c063d0979112f4b9f4eeace04c8cf5ff/docs/projects#implementation-plans-status-in-rfc -->
+<!-- This template is exhaustive and may include sections which aren't relevant to your project. Feel free to remove any sections which would not be useful to have. -->
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @sarayourfriend
+- [ ] @zackkrida
+
+## Project links
+
+<!-- Enumerate any references to other documents/pages, including milestones and other plans -->
+
+- [Project Thread](https://github.com/WordPress/openverse/issues/411)
+- [Project Proposal](https://docs.openverse.org/projects/proposals/nuxt_3_migration/20230604-project_proposal_nuxt_3_migration.html)
+
+## Overview
+
+<!-- An overview of the implementation plan, if necessary. Save any specific steps for the section(s) below. -->
+
+Nuxt team has created Nuxt bridge to facilitate the migration from Nuxt 2.
+However, there seems to be
+[no support for Nuxt Bridge in Nuxt i18n](https://github.com/nuxt-modules/i18n/discussions/1334#discussioncomment-1967220),
+so the best way forward is to migrate directly to Nuxt 3.
+
+Nuxt 3 migration will require a lot of changes split into several PRs.
+
+One set of changes is required for Nuxt 3, but can also work in the current Nuxt
+2 app. An example of such change is the conversion of `onBeforeRouteEnter` to
+`middleware`: both work well in Nuxt 2, but only `middleware` can access `store`
+in Nuxt 3. These changes are described in the "Changes to main" section below,
+should be merged directly into `main` branch.
+
+The other changes like actually updating the `nuxt` version would break the
+current app. To make sure that we can split these changes into several distinct
+steps without breaking the production app, we should use a separate branch,
+`nuxt3`, that would gradually get more and more features re-enabled.
+
+### Nuxt 3 / Vue 3 features in Openverse: discussion
+
+#### Autoimports
+
+Nuxt 3 auto imports composables and components. Unlike the usual global imports,
+these imports are handled in such a way that only the imports that are actually
+used are left in the production files. The Nuxt team have optimized the way they
+work both in the working app and in the IDEs, so I think we _should_ use the
+auto import feature and not go against the framework and disable it. In the
+current setup, our test suite throws warnings when a child component is not
+imported. We would need to make sure that this does not happen with Nuxt 3.
```

Comment:
I anticipated a similar response :) Thank you for describing the reasons in such a detail, it's really helpful.

On the first point, I looked at the Nuxt 3 draft branch locally to see if the IDE support helps with it. And even though I can click "Go to definition" on a non-imported component and open the component file, I can't do the same for a composable. Going to definition there sends you to the file that lists all of the auto-imported items, so while it _is_ possible to find the source of the composable, it's not as straightforward as when we import it. 
> Variable name collision and shadowing is also much harder to understand or anticipate side effects for.
This is also a serious consideration.

The second point - I guess you can't understand the extent of how bad it can be until that happens to you :)
I agree with the third point, too.

What I wanted to avoid was going against the framework. This would cause us to have to add work-arounds around the magic, which might need a lot of work. However, seeing that the IDE support is not as great as I was expecting, and considering all of the drawbacks of using auto-imports, I'm open to _not_ using this feature.

Reply:
After writing this comment, I decided to check if we would have problems with the imports and tests. In the current Nuxt 2 app, the components are auto-imported, so the app can run even if we don't add all component imports. However, the tests show warnings for components that weren't imported.

In my first try, this was a big problem with explicitly adding all of the imports. The biggest problem is the imports from `.nuxt` and the imports that start with `#`, especially deep within modules (`nuxtjs/svg-sprite` being the one I tried to work around the most). Not saying that this is a blocker, but just noting here that getting all of the imports to work both in the prod app and the tests would require a lot of debugging.

Reply:
> And even though I can click "Go to definition" on a non-imported component and open the component file, I can't do the same for a composable. Going to definition there sends you to the file that lists all of the auto-imported items, so while it is possible to find the source of the composable, it's not as straightforward as when we import it.

This reminds me actually, that Nuxt 3's auto-imports I believe is the main reason why running TypeScript _requires_ a build. Nuxt 3's build process pulls the auto-imported composables out and generates the tsconfig.json to include all the ambient types that you're meant to extend in your Nuxt app's tsconfig.json.

I wonder how much all those ambient types slow down TypeScript (or perhaps increase its memory footprint).

In either case, we'd need to change CI to run the types check only after a Nuxt 3 build is run, at least if that process is still the same, which sounds to be the case based on the IDE behaviour of go-to-definition for the composables.

> Not saying that this is a blocker, but just noting here that getting all of the imports to work both in the prod app and the tests would require a lot of debugging

That's a shame :slightly_frowning_face:. When I tried to convert Nuxt using bridge, the option to disable auto-imports didn't even work. Does turning off auto-imports stop component auto-importing in the app build? The Nuxt 3 only reference auto-importing for _composables_, both in the composables documentation and in the auto-imports documentation :thinking:

Reply:
> In either case, we'd need to change CI to run the types check only after a Nuxt 3 build is run, at least if that process is still the same, which sounds to be the case based on the IDE behaviour of go-to-definition for the composables.

I will add this to the proposal, thanks! (Check if we need to change CI to run the types check after a Nuxt 3 build is run - change if necessary)

> The Nuxt 3 only reference auto-importing for composables, both in the composables documentation and in the auto-imports documentation 🤔
I'm not sure if I understand this sentence correctly, but Nuxt 3 docs do say that they auto import components, composables and even utils (although the way it's worded is not very straightforward): https://nuxt.com/docs/guide/concepts/auto-imports#directory-based-auto-imports

By the way, I didn't realize that the recommended way to explicitly add imports is by using `#imports`, found while re-reading the docs for this comment.

Reply:
> Check if we need to change CI to run the types check after a Nuxt 3 build is run - change if necessary

This would also need to change in the pre-commit, which is probably the place it will hurt the DevEx the most.

Reply:
I definitely understand the desire to do what's idiomatic for the framework _and_ absolutely hating this automagical import stuff. 

> Does turning off auto-imports stop component auto-importing in the app build? 

I just tested this with a fresh Nuxt 3 install. Sadly, components still appear to be auto imported even with this config:

```ts
export default defineNuxtConfig({
  imports: {
    autoImport: false
  }
})
```

It _does_ work for composeables, but not components. 

I'm really conflicted on how to proceed for this one.

Reply:
Actually, if it auto-imports components, I think that's fine. That's the current behaviour of Nuxt 2, so it would present effectively zero change to our workflow.

Reply:
Oh, true!